mlflow/pyspark/ml/_autolog.py

```diff
@@ -53,43 +53,25 @@ def _get_struct_type_by_cols(input_fields: Set[str], df_schema: t.StructType) ->
 
 
 def get_feature_cols(
-    df_schema: t.StructType,
+    df: DataFrame,
     transformer: Union[Transformer, PipelineModel],
-    input_fields: Set[str] = None,
-    spark: SparkSession = None,
 ) -> Set[str]:
     """
     Finds feature columns from an input dataset. If a dataset
     contains non-feature columns, those columns are not returned, but
     if `input_fields` is set to include non-feature columns those
     will be included in the return set of column names.
 
-    :param df_schema: An input spark schema to look for the feature columns
+    :param df: An input spark dataframe.
     :param transformer: A pipeline/transformer to get the required feature columns
-    :param input_fields: Initial columns to keep in the returned list
-    :param spark: A spark session
     :return: A set of all the feature columns that are required
              for the pipeline/transformer plus any initial columns passed in.
     """
-    if spark is None:
-        spark = SparkSession.builder.getOrCreate()
-    if input_fields is None:
-        input_fields = set()
-        # Using an empty dataset doesn't work for all estimators
-        # such as: pyspark.ml.classification.OneVsRest, so set
-        # a single row and column of double type
-        df = spark.createDataFrame([(1.0,)])
-    else:
-        df = spark.createDataFrame(
-            spark.sparkContext.emptyRDD(),
-            _get_struct_type_by_cols(input_fields, df_schema),
-        )
-    try:
-        _do_pipeline_transform(df, transformer)
-    except IllegalArgumentException as iae:
-        col_name_search = re.search("(.*) does not exist.", iae.desc, re.IGNORECASE)
-        if col_name_search:
-            col_name = col_name_search.group(1)
-            input_fields.add(col_name)
-            get_feature_cols(df_schema, transformer, input_fields, spark)
-    return input_fields
+    feature_cols = set()
+    for column in df.columns:
+        try:
+            transformer.transform(df.drop(column).limit(1))
+        except IllegalArgumentException as iae:
+            if re.search("(.*) does not exist.", iae.desc, re.IGNORECASE):
```

Comment:
What should we do if exception raised but error message does not match "(.*) does not exist" ?

Reply:
I think we can throw an exception.

Reply:
```suggestion
            feature_cols = {}
            df_subset = df.limit(1).cache()
            for column in df_subset.columns:
              ...
            df_subset.unpersist()
```

Reply:
We might need add a check and raise specific exception if the `df_subset` contains zero rows.

Reply:
When does `df_subset` have zero rows?